locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/softlayer/SoftLayerSameVlanLocationCustomizer.java

```diff
@@ -0,0 +1,192 @@
+import org.apache.brooklyn.location.jclouds.JcloudsLocation;
+import org.apache.brooklyn.location.jclouds.JcloudsMachineLocation;
+import org.apache.brooklyn.util.core.task.Tasks;
+import org.apache.brooklyn.util.javalang.Threads;
+import org.apache.brooklyn.util.text.Strings;
+import org.apache.brooklyn.util.time.Duration;
+
+/**
+ * Configures SoftLayer provisioned VMs with the same VLAN ids across an application.
+ */
+public class SoftLayerSameVlanLocationCustomizer extends BasicJcloudsLocationCustomizer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SoftLayerSameVlanLocationCustomizer.class);
+
+    public static final ConfigKey<String> SCOPE_UID = ConfigKeys.newStringConfigKey("softlayer.vlan.scopeUid",
+            "The unique identifier for a Softlayer location scope that will have VMs created in the same VLAN");
+
+    public static final ConfigKey<Duration> SCOPE_TIMEOUT = ConfigKeys.newDurationConfigKey("softlayer.vlan.timeout",
+            "The length of time to wait for a Softlayer VLAN ID", Duration.minutes(10));
+
+    /** Caches the semaphore object for locking access to the scope. */
+    private static final Cache<String, Semaphore> semaphoreCache = CacheBuilder.newBuilder().build();
+
+    /** Caches the public VLAN id that should be shared between all entities in the same scope. */
+    private static final Cache<String, Integer> publicVlanIdCache = CacheBuilder.newBuilder().build();
+
+    /** Caches the private VLAN id that should be shared between all entities in the same scope. */
+    private static final Cache<String, Integer> privateVlanIdCache = CacheBuilder.newBuilder().build();
+
+    /**
+     * Used to obtain the VLANs being used by the first created {@link JcloudsMachineLocation}.
+     */
+    @Override
+    public void customize(JcloudsLocation location, ComputeService computeService, JcloudsMachineLocation machine) {
+        String provider = location.getProvider();
+        if (!(provider.equals(SoftLayerConstants.SOFTLAYER_PROVIDER_NAME))) {
+            String message = String.format("Invalid location provider: %s", provider);
+            LOG.warn(message);
+            throw new IllegalArgumentException(message);
+        }
+
+        Semaphore semaphore = null;
+        String scopeUid = getScopeUid(location);
+
+        synchronized (SoftLayerSameVlanLocationCustomizer.class) {
+            semaphore = semaphoreCache.getIfPresent(scopeUid);
+            if (semaphore == null) {
+                throw new IllegalStateException("No semaphore available for scope: " + scopeUid);
+            }
+        }
+
+        Integer publicVlanId = publicVlanIdCache.getIfPresent(scopeUid);
+        Integer privateVlanId = privateVlanIdCache.getIfPresent(scopeUid);
+        if (privateVlanId != null || publicVlanId != null) {
+            LOG.debug("SoftLayer VLANs already configured for application: {}", scopeUid);
+            return;
+        }
+
```

Comment:
add comment

    // no vlan info yet known, we are probably the vm and we should set the vlan info for others to then learn about